sympy/matrices/matrices.py

```diff
@@ -1538,19 +1538,35 @@ def eig_mat(val, pow):
             return mat_cache[(val, pow)]
 
         # helper functions
-        def nullity_chain(val):
+        def nullity_chain(val, algebraic_multiplicity):
             """Calculate the sequence  [0, nullity(E), nullity(E**2), ...]
             until it is constant where ``E = self - val*I``"""
             # mat.rank() is faster than computing the null space,
             # so use the rank-nullity theorem
             cols = self.cols
             ret = [0]
-            nullity = cols - eig_mat(val, 1).rank()
-            i = 2
-            while nullity != ret[-1]:
-                ret.append(nullity)
-                nullity = cols - eig_mat(val, i).rank()
-                i += 1
+
+            nullity_prev = 0
+            i = 1
+            while True:
+                nullity_next = cols - eig_mat(val, i).rank()
+
+                monotonic_check = nullity_next == nullity_prev
+                multiplicity_check = nullity_prev == algebraic_multiplicity
+
+                if monotonic_check is True and multiplicity_check is True:
+                    break
+                elif monotonic_check is False and multiplicity_check is False:
```

Comment:
I feel like this is getting confusing and unnecessarily complicated. Can you revert this to the old function but add an early bail when the multiplicity is reached and monotonicity check?